lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -1005,23 +1008,37 @@ func (d *zfs) createVolumeFromMigrationOptimized(vol Volume, conn io.ReadWriteCl
 
 		// Transfer the snapshots.
 		for _, snapName := range volTargetArgs.Snapshots {
-			fullSnapshotName := GetSnapshotVolumeName(vol.name, snapName)
-			wrapper := migration.ProgressWriter(op, "fs_progress", fullSnapshotName)
-
-			err = d.receiveDataset(vol, conn, wrapper)
+			snapVol, err := vol.NewSnapshot(snapName)
 			if err != nil {
 				return err
 			}
+
+			wrapper := migration.ProgressWriter(op, "fs_progress", snapVol.Name())
+
+			err = d.receiveDataset(snapVol, conn, wrapper)
```

Comment:
@stgraber @monstermunchkin I'm now passing the full desired snaphot volume in her rather than what we were doing previously which was passing the parent volume. According to the `zfs recieve` manpage this is fine to do (and it appears to work fines). If you pass in the parent name it seems to use the name of the snapshot being sent to create the new snapshot, but if you pass it in explicitly it uses that instead. In our case they should be the same in any case and helps with readability to be passing in the snapshot volume rather than the parent.